core/circuit_breaker/rule_manager.go

```diff
@@ -5,31 +5,115 @@ import (
 	"strings"
 	"sync"
 
+	"github.com/alibaba/sentinel-golang/core/stat/base"
 	"github.com/alibaba/sentinel-golang/logging"
 )
 
-var (
-	logger = logging.GetDefaultLogger()
+const (
+	CommonRulesResource = "__circuitBreakerCommonRulesResource__"
 )
 
+type CircuitBreakerGenFunc func(Rule) CircuitBreaker
+
 var (
-	breakerRules = make(map[string][]Rule)
-	breakers     = make(map[string][]CircuitBreaker)
-	ruleMux      = &sync.RWMutex{}
+	logger                = logging.GetDefaultLogger()
+	breakerRules          = make(map[string][]Rule)
+	cbGenFuncMap          = make(map[Strategy]CircuitBreakerGenFunc)
+	breakers              = make(map[string][]CircuitBreaker)
+	updateMux             = &sync.RWMutex{}
+	statusSwitchListeners = make([]StatusSwitchListener, 0)
 )
 
+func init() {
+	cbGenFuncMap[SlowRt] = func(r Rule) CircuitBreaker {
+		rtRule := r.(*slowRtRule)
+		if rtRule == nil {
+			return nil
+		}
+		status := new(State)
+		status.set(Closed)
+
+		interval := rtRule.StatIntervalMs
+		stat := &slowRequestLeapArray{}
+		stat.data = base.NewLeapArray(1, interval, stat)
+
+		return &slowRtCircuitBreaker{
+			circuitBreakerBase: circuitBreakerBase{
+				rule:                  rtRule,
+				retryTimeoutMs:        rtRule.RetryTimeout,
+				nextRetryTimestamp:    0,
+				status:                status,
+				statusSwitchListeners: statusSwitchListeners,
+			},
+			stat:                stat,
+			maxAllowedRt:        rtRule.MaxAllowedRt,
+			maxSlowRequestRatio: rtRule.MaxSlowRequestRatio,
+			minRequestAmount:    rtRule.MinRequestAmount,
+		}
+	}
+	cbGenFuncMap[ErrorRatio] = func(r Rule) CircuitBreaker {
+		errRatioRule := r.(*errorRatioRule)
+		if errRatioRule == nil {
+			return nil
+		}
+		status := new(State)
+		status.set(Closed)
+
+		interval := errRatioRule.StatIntervalMs
+		stat := &errorCounterLeapArray{}
+		stat.data = base.NewLeapArray(1, interval, stat)
+
+		return &errorRatioCircuitBreaker{
+			circuitBreakerBase: circuitBreakerBase{
+				rule:                  errRatioRule,
+				retryTimeoutMs:        errRatioRule.RetryTimeout,
+				nextRetryTimestamp:    0,
+				status:                status,
+				statusSwitchListeners: statusSwitchListeners,
```

Comment:
The new listeners will not be included if added after the circuit breaker is generated.